lib/entities/SubscriptionSet.js

```diff
@@ -192,11 +170,36 @@ class SubscriptionSet extends SubscribeCapable_1.SubscribeCapable {
      * @internal
      */
     updateListeners() {
-        if (!this.aggregatedListener)
-            return;
+        // Actual list of channels and groups.
+        const channelGroups = [];
+        const channels = [];
+        // Gather actual information about active channels / groups.
+        this.subscriptionList.forEach((subscription) => {
+            if (subscription.channelGroups.length)
+                channelGroups.push(...subscription.channelGroups);
+            if (subscription.channels.length)
+                channels.push(...subscription.channels);
+        });
+        // Identify channels / groups which should be added / removed.
+        const channelsToRemove = this.channelNames.filter((channel) => !channels.includes(channel));
+        const groupsToRemove = this.groupNames.filter((group) => !channelGroups.includes(group));
+        const channelsToAdd = channels.filter((channel) => !this.channelNames.includes(channel));
+        const groupsToAdd = channelGroups.filter((group) => !this.groupNames.includes(group));
+        // Removing type-based listener for unused channels / groups.
+        if (channelsToRemove.length || groupsToRemove.length) {
+            this.eventEmitter.removeListener(this.typeBasedListener, this.typeBasedListenerId, channelsToRemove, groupsToRemove);
+        }
+        // Adding type-based listener for unused channels / groups.
+        if (channelsToAdd.length || groupsToAdd.length)
+            this.eventEmitter.addListener(this.typeBasedListener, channelsToAdd, groupsToAdd, this.typeBasedListenerId);
         const aggregatedListener = this.aggregatedListener;
-        this.removeListener(this.aggregatedListener);
-        this.addListener(aggregatedListener);
+        if (aggregatedListener)
```

Comment:
Great to see this optimisation for updating listeners. Looks more readable and understandable than previous complex logic of finding the entity names to be removed. 👍🏻